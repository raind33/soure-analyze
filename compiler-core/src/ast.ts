export enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  SIMPLE_ELEMENT,
  TEXT
}