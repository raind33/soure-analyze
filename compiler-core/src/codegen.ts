export function generate(ast:any) {
  const context = createCodegenContext()
  const args = ['_ctx', '_cache', '$props', '$setup', '$data', '$options']
  const functionName = 'render'
  const renderContent = `return "rain"`
  context.push(`function ${functionName}(${args.join(',')}) {`)
  context.push(renderContent)
  return context.code
}

function createCodegenContext() {
  const obj = {
    code: '',
    push(str: any) {
      obj.code+=str
    }
  }
  return obj
}