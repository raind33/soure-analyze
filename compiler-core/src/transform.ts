export function transform(root, options) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
}

function traverseNode(node:any, context:any) {
  context.nodeTransforms.forEach(transform => {
    transform(node)
  })
  transformChildren(node, context)
}
function transformChildren(node:any, context:any) {
  if(node.children) {
    node.children.forEach(node => {
      traverseNode(node, context)
    })
  }
}
function createTransformContext(root:any, options:any) {
  return {
    root,
    nodeTransforms: options.nodeTransforms || []
  }
}