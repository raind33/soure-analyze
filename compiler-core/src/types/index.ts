export interface Context {
  source: string
}