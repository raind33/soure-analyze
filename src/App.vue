<script setup lang="ts">
import { useStore } from 'vuex4';

const store = useStore()
console.log(store)
const add = () => {
  store.commit('user/setA', 999)
}
</script>

<template>
  <img alt="Vue logo" src="./assets/logo.png" />
  {{store.getters['user/b']}}
  <button @click="add">点击</button>
</template>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
  margin-top: 60px;
}
</style>
