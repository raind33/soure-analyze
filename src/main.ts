import { createApp } from '../runtime-core'
// import App from './App.vue'
import TestVue from './test-vue.vue'
// import store from '@/store'

createApp(TestVue).mount(document.querySelector('#app'))
