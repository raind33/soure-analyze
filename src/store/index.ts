import { createStore } from '../../vuex4'
import user from './user'
import test from './testStore'

export default createStore({
  state: {
    a: 1
  },
  getters: {
    b(state) {
      return state.a
    }
  },
  mutations: {
    setA(state, payload) {
      state.a = payload
    }
  },
  actions: {
    getData( { commit }) {
      commit('setA')
    }
  },
  modules: {
    user,
    test
  }
})