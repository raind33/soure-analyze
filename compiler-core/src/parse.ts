import { NodeTypes } from "./ast"
import { Context } from "./types"

enum TagType {
  START,
  END
}
export function baseParse(content:string) {
  const context = createParseContext(content)
  return createRoot(parseChildren(context))
}

function createParseContext(content:string) {
  return {
    source: content
  }
}
function createRoot(children: any) {
  return {
    children
  }
}

function parseChildren(context:Context) {
  const nodes:any = []
  const s = context.source
  let node
  if(s.startsWith('{{')){
    node = parseInterpolation(context)
    nodes.push(node)
  } else if(s.startsWith('<')){
    if(/[a-z]/i.test(s[1])) {
      node = parseElement(context)
    }
  }
  nodes.push(node)

  return nodes
}
function parseElement(context:Context) {
  const element = parseTag(context, TagType.START)
  parseTag(context, TagType.END)
  return element
}
function parseTag(context:Context, type: TagType) {
  const match = /^<\/?([a-z]*)/i.exec(context.source)
  const content = match[1]
  advance(context, match[0].length)
  advance(context, 1)
  if(type === TagType.END) return
  return {
    type: NodeTypes.ELEMENT,
    tag: content
  }
}
function parseInterpolation(context:Context) {
  const closeDelimiter = '}}'
  const openDelimiter = '{{'
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
  const rawContentlength = closeIndex - openDelimiter.length
  advance(context, openDelimiter.length)
  const rawContent = context.source.slice(0, rawContentlength)
  const content = rawContent.trim()
  advance(context, rawContentlength+closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    }
  }
}
function advance(context:Context, length:number) {
  context.source = context.source.slice(length)
}