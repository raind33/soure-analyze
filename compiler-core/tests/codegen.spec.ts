import { NodeTypes } from "../src/ast"
import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"

describe('generate', () => {
  test('text', () => {
    const ast = baseParse('rain')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot()
  })
})