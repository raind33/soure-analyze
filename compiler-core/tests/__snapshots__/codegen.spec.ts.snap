// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate text 1`] = `"function render(_ctx,_cache,$props,$setup,$data,$options) {return \\"rain\\""`;
